### PR TITLE
Save Buffers and FDs in RecoverySourceHandler

### PR DESCRIPTION
**Pull request by original-brownbear**
Using the same trick here that we use in snapshots.
Segment and segment infos files are on heap here in most cases so
no need to allocate a buffer for them or do file IO.
Also, we can size the 0.5M buffer smaller when we know the largest file to
be less than 0.5M.

Both of these changes are probably of trivial effect in production in most cases
but may save quite a bit of GC noise in tests where we run small heaps and repeated
large allocations of these buffers might be a source of the random GC slowness
we see here and there.

### CODE REVIEW

**Review by fcofdez — APPROVED** [review]
LGTM 👍

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 1]
```diff
@@ -903,14 +905,23 @@ void sendFiles(Store store, StoreFileMetadata[] files, IntSupplier translogOps,
...
+                        }
                     }
 
                     @Override
                     protected FileChunk nextChunkRequest(StoreFileMetadata md) throws IOException {
                         assert Transports.assertNotTransportThread("read file chunk");
                         cancellableThreads.checkForCancel();
-                        final byte[] buffer = Objects.requireNonNullElseGet(buffers.pollFirst(), () -> new byte[chunkSizeInBytes]);
+                        if (currentInput == null) {
+                            // no input => reading directly from the metadata
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:1128]: ++ added :)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/packaging-tests-unix-sample

**Comment by original-brownbear** [comment]
Thanks Francisco!

**Pull request merged**